# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
# Commit: e44f2257d1ad
# Full Hash: e44f2257d1ad95875d36b664109e9c15f62e8d2e
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Set keyframe request method in VideoConduit, based on negotiation. r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D105775
# ==============================================================================

diff -r 7330681cf4de -r e44f2257d1ad dom/media/webrtc/libwebrtcglue/VideoConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Fri Feb 19 13:45:49 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/VideoConduit.cpp	Fri Feb 19 13:47:26 2021 +0100
@@ -1099,6 +1099,8 @@
     return kMediaConduitMalformedArgument;
   }
 
+  webrtc::KeyFrameReqMethod kf_request_method =
+      webrtc::KeyFrameReqMethod::kNone;
   bool use_nack_basic = false;
   bool use_tmmbr = false;
   bool use_fec = false;
@@ -1137,6 +1139,14 @@
       continue;
     }
 
+    // Check for the keyframe request type: PLI is preferred over FIR, and FIR
+    // is preferred over none.
+    if (codec_config->RtcpFbNackIsSet("pli")) {
+      kf_request_method = webrtc::KeyFrameReqMethod::kPliRtcp;
+    } else if (codec_config->RtcpFbCcmIsSet("fir")) {
+      kf_request_method = webrtc::KeyFrameReqMethod::kFirRtcp;
+    }
+
     // What if codec A has Nack and REMB, and codec B has TMMBR, and codec C has
     // none? In practice, that's not a useful configuration, and
     // VideoReceiveStream::Config can't represent that, so simply union the
@@ -1160,6 +1170,7 @@
           (use_nack_basic ? 1000 : 0) ||
       mRecvStreamConfig.rtp.transport_cc != use_transport_cc ||
       mRecvStreamConfig.rtp.tmmbr != use_tmmbr ||
+      mRecvStreamConfig.rtp.keyframe_method != kf_request_method ||
       (use_fec &&
        (mRecvStreamConfig.rtp.ulpfec_payload_type != ulpfec_payload_type ||
         mRecvStreamConfig.rtp.red_payload_type != red_payload_type))) {
@@ -1176,7 +1187,7 @@
     mRecvStreamConfig.rtp.nack.rtp_history_ms = use_nack_basic ? 1000 : 0;
     mRecvStreamConfig.rtp.transport_cc = use_transport_cc;
     mRecvStreamConfig.rtp.tmmbr = use_tmmbr;
-    // mRecvStreamConfig.rtp.keyframe_method = kf_request_method;
+    mRecvStreamConfig.rtp.keyframe_method = kf_request_method;
 
     if (use_fec) {
       mRecvStreamConfig.rtp.ulpfec_payload_type = ulpfec_payload_type;
